apps/emqx_rule_engine/src/emqx_rule_engine_api.erl

```diff
@@ -423,7 +423,12 @@ param_path_id() ->
 %% Internal functions
 %%------------------------------------------------------------------------------
 
-err_msg(Msg) -> emqx_misc:readable_error_msg(Msg).
+err_msg({RuleError, {_E, R, _S}}) when is_tuple(R) ->
+    emqx_misc:readable_error_msg(emqx_json:encode([{RuleError, element(1, R)}]));
```

Comment:
just question, what if encode crashes?

Reply:
Yeah, good question, I went with "let's just risk it" in this case. It's more of an issue with line 429 though. I can wrap it in a try catch and default to the original behavior.